Additional function arg unpacking (PEP 448)

dlech opened the pull request:
This is an attempt to implement PEP 448 for function arguments. It allows multiple *args and **args to be unpacked. (Popular request in #1329).

I'm curious to see how it affects code size and performance.

There are still some todo items:
- There is still a null object emitted in the bytecodes that used to be the **args that needs to be removed
- There are probably edge cases where memory allocation is wrong, especially in the case where one or more of the *args are iterators instead of list/tuple.

dlech committed:
py/runtime: Allow multiple **args in function call.

This is a partial implementation of PEP 448 to allow multiple **
unpackings when calling a function or method.

The compiler is modified to encode the argument as a None: obj key-value
pair (similar to how regular keyword arguments are encoded as str: obj
pairs). The extra object that was pushed on the stack to hold a single
** unpacking object is no longer used and is removed.

The runtime is modified to decode this new format.

Signed-off-by: David Lechner <david@pybricks.com>

dlech committed:
py/runtime: Allow multiple *args in function call.

This is a partial implementation of PEP 448 to allow unpacking multiple
star args in a function or method call.

This is implemented by changing the emitted bytecodes so that both
positional args and star args are stored as positional args. A bitmap
is added to indicate if an argument at a given position is a positional
argument or a star arg.

In the bytecodes, this new bitmap takes the place of the old star arg.
It is stored as a small int, so this means only the first N arguments
can be star args where N is the number of bits in a small int.

The runtime is modified to interpret this new bytecode format while
still trying to perform as few memory reallocations as possible.

Signed-off-by: David Lechner <david@pybricks.com>

tve commented:
might want to fix the typo in the title ;-)

dlech commented:
If that is my worst mistake, we are in good shape 😆

dpgeorge commented:
Thanks for working on this.  It's a tough problem, especially to make it efficient (in RAM usage, execution time, code size).  Properly reviewing this would take some time.

For the record, the code size change is:
```
   bare-arm:   +48 +0.072% 
minimal x86:  +128 +0.086% 
   unix x64:  +144 +0.029% 
unix nanbox:  +272 +0.061% 
      stm32:   +52 +0.013% PYBV10
     cc3200:   +48 +0.026% 
    esp8266:   +72 +0.011% GENERIC
      esp32:   +36 +0.003% GENERIC[incl -32(data)]
        nrf:   +52 +0.036% pca10040
       samd:   +48 +0.047% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

dpgeorge commented:
Can you please rebase this on latest master?

dlech commented:
Rebased.

dlech commented:
> * There is still a null object emitted in the bytecodes that used to be the **args that needs to be removed

Added a commit to address this todo item.

dpgeorge commented on py/compile.c:
```diff
@@ -2397,24 +2397,25 @@ STATIC void compile_trailer_paren_helper(compiler_t *comp, mp_parse_node_t pn_ar
                 }
                 star_flags |= MP_EMIT_STAR_FLAG_SINGLE;
-                star_args_node = pns_arg;
+                star_args |= 1 << i;
```
It'd probably be a good idea to check that this still fits in a small int, and raise an exception if not.

dlech replied:
I'm not sure this was worth the extra code size since it is rare that anyone would use that many arguments, but I added it anyway.

dpgeorge commented on tests/basics/python34.py:
```diff
@@ -6,27 +6,23 @@
 def print_ret(x):
     print(x)
     return x
-f4(*print_ret(['a', 'b']), kw_arg=print_ret(None))
```
I think this test should be retained, it's a good test for evaluation order.  Maybe put it back in `basics/fun_kwvarargs.py` now that both uPy and CPy agree on the output?

dlech replied:
done.

dpgeorge commented:
>> There is still a null object emitted in the bytecodes that used to be the **args that needs to be removed
>
> Added a commit to address this todo item.

IMO it makes sense to squash this commit into the first one, so that the `cmd_showbc` test does not change so much and then get reverted.  That would make it a bit clearer as to the impact of this change, and how it changes generated bytecode.

> There are probably edge cases where memory allocation is wrong, especially in the case where one or more of the *args are iterators instead of list/tuple.

Do you mean it might segfault in an edge case?  Or that it simply makes a bad guess at how much memory to preallocate?  As far as I can tell it all looks OK.

-----------

In fact this has exposed a bug in current master, the following will segfault:
```
def f(*a, **b):
    print(a)
    print(b)
f(*range(100), **{str(i):i for i in range(100)})
```
With this PR, that bug is fixed.

dpgeorge commented:
I think the general approach here is very clever.  When `*seq` or `**dict` is used as an argument the Python stack looks like this:
```
fun(arg0, arg1, *seq, key0=value0, key1=value1, **dict)
old: fun arg0 arg1 key0 value0 key1 value1 seq dict
new: fun arg0 arg1 seq key0 value0 key1 value1 NULL dict 4
```
For simple cases:
```
fun(*seq)
old: fun seq NULL
new: fun seq 1
```
```
fun(**dict)
old: fun NULL dict
new: fun NULL dict 0
```
Also in the case of using `**dict` the bytecode for the opcode increased by 1 byte because in now needs 2 bytes to encode the `unum = (npos | nkw <<8)` value.  As such `fun(**dict)` uses 2 extra bytes in total in the bytecode (one for the unum, one for the seq bitmap).

I think this can be optimised.  Eg if npos==0 then don't store the bitmap.  And if npos==0 and nkw==0 then implicitly add 1 to nkw.

dlech commented:
> IMO it makes sense to squash this commit into the first one,

Agreed.

> Do you mean it might segfault in an edge case? Or that it simply makes a bad guess at how much memory to preallocate?

The latter.

> I think this can be optimised. Eg if npos==0 then don't store the bitmap. And if npos==0 and nkw==0 then implicitly add 1 to nkw.

I will have a look.

dlech commented:
I have updated this with the suggested changes. With regard to the suggested optimizations, I did it a bit differently. " if npos==0 then don't store the bitmap." was simple enough but "And if npos==0 and nkw==0 then implicitly add 1 to nkw." got messy since it requires checking for this in many places. So, I figured that if we were going to all all of these checks anyway, we could do it slightly differently and optimize (at least in terms of fewer bytes emitted) more cases. See the commit message for the full explanation.

dpgeorge commented:
> Given that arg unpacking is probably not frequently used, I'm not sure the trade off in compiler size/complexity is worth the savings in emitted byte code size.

You might be right, it does look complicated.  I will review it.

peterhinch commented:
FWIW I think this is a very useful feature - subject to code size, of course :)

For example in GUI code, widgets may have multiple constructor args to define colors, shapes, fonts, sizes, callbacks etc. A `dict` enables a set of common attributes to be assigned to multiple widgets. The ability to assign more than one `dict` would be useful. You might have one to define attributes common to all controls and another for additional attributes applicable only to (say) `Meter` instances.

dpgeorge commented on tests/basics/fun_kwargs.py:
```diff
@@ -33,3 +33,15 @@ def f3(a, b, *args):
+    return x
+
+
+f4(*print_ret(["a", "b"]), kw_arg=print_ret(None))
```
this test should go in `tests/baics/fun_kwvarargs.py`

dlech replied:
done

dpgeorge commented:
Code size change for supporting multiple `*` and `**` args is:
```
   bare-arm:   +80 +0.139% 
minimal x86:  +133 +0.081% 
   unix x64:  +144 +0.028% 
unix nanbox:  +300 +0.065% 
      stm32:   +68 +0.017% PYBV10
     cc3200:   +80 +0.044% 
    esp8266:   +92 +0.013% GENERIC
      esp32:   +92 +0.006% GENERIC[incl -16(data)]
        nrf:   +16 +0.009% pca10040
        rp2:   +48 +0.010% PICO
       samd:   +64 +0.046% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
Then the additional code size change on top of that for the encoding optimisation is:
```
   bare-arm:  +120 +0.208% 
minimal x86:  +204 +0.124% 
   unix x64:  +374 +0.071% 
unix nanbox:  +592 +0.128% 
      stm32:  +160 +0.041% PYBV10
     cc3200:  +128 +0.070% 
    esp8266:  +156 +0.022% GENERIC
      esp32:  +188 +0.012% GENERIC
        nrf:  +148 +0.085% pca10040
        rp2:  +136 +0.028% PICO
       samd:   -56 -0.040% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
I'd say that optimisation is not worth it.  It's relatively rare to use `*` / `**` args and there are better ways than that to spend bytes on reducing/optimising the bytecode.

So please remove that encoding optimisation commit from this PR.  That should also fix the CI issues with native emitter.

dpgeorge commented on py/vm.c:
```diff
@@ -927,8 +927,8 @@ unwind_jump:;
                     // We have following stack layout here:
-                    // fun arg0 arg1 ... kw0 val0 kw1 val1 ... seq dict <- TOS
-                    sp -= (unum & 0xff) + ((unum >> 7) & 0x1fe) + 2;
+                    // fun arg0 arg1 ... kw0 val0 kw1 val1 ... bitmap <- TOS
```
In this commit the `bitmap` does not exist so it should be `seq` instead.  I.e., this commit should just delete the `dict` entry.  Then in the next commit, `seq` is renamed to `bitmap`.

dlech replied:
done

dpgeorge commented:
Performance change (without the encoding optimisation commit), on PYBv1.0:
```
diff of scores (higher is better)
N=100 M=100                        p0 ->         p1         diff      diff% (error%)
bm_chaos.py                    351.40 ->     355.38 :      +3.98 =  +1.133% (+/-0.01%)
bm_fannkuch.py                  77.31 ->      77.98 :      +0.67 =  +0.867% (+/-0.01%)
bm_fft.py                     2526.30 ->    2549.84 :     +23.54 =  +0.932% (+/-0.00%)
bm_float.py                   5905.82 ->    5939.79 :     +33.97 =  +0.575% (+/-0.00%)
bm_hexiom.py                    47.17 ->      47.27 :      +0.10 =  +0.212% (+/-0.01%)
bm_nqueens.py                 4419.24 ->    4451.92 :     +32.68 =  +0.739% (+/-0.00%)
bm_pidigits.py                 638.01 ->     640.70 :      +2.69 =  +0.422% (+/-0.23%)
core_import_mpy_multi.py       553.34 ->     556.64 :      +3.30 =  +0.596% (+/-0.01%)
core_import_mpy_single.py       68.36 ->      68.69 :      +0.33 =  +0.483% (+/-0.01%)
core_qstr.py                    63.73 ->      63.86 :      +0.13 =  +0.204% (+/-0.01%)
core_yield_from.py             356.12 ->     355.00 :      -1.12 =  -0.315% (+/-0.00%)
misc_aes.py                    411.90 ->     413.86 :      +1.96 =  +0.476% (+/-0.01%)
misc_mandel.py                3350.85 ->    3414.97 :     +64.12 =  +1.914% (+/-0.00%)
misc_pystone.py               2225.86 ->    2232.44 :      +6.58 =  +0.296% (+/-0.00%)
misc_raytrace.py               368.11 ->     372.18 :      +4.07 =  +1.106% (+/-0.00%)
viper_call0.py                 572.83 ->     572.73 :      -0.10 =  -0.017% (+/-0.00%)
viper_call1a.py                546.71 ->     546.71 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1b.py                436.01 ->     435.91 :      -0.10 =  -0.023% (+/-0.01%)
viper_call1c.py                440.59 ->     440.51 :      -0.08 =  -0.018% (+/-0.00%)
viper_call2a.py                532.82 ->     532.85 :      +0.03 =  +0.006% (+/-0.01%)
viper_call2b.py                377.24 ->     377.17 :      -0.07 =  -0.019% (+/-0.00%)

```

dlech commented:
The changes in code coverage show some potential issues: 

- [this branch](https://app.codecov.io/gh/micropython/micropython/compare/5807/diff#D3L2418) is not hit because the cpydiff isn't run in the coverage tests.
- [this branch](https://app.codecov.io/gh/micropython/micropython/compare/5807/diff#D4L816) and [this branch](https://app.codecov.io/gh/micropython/micropython/compare/5807/diff#D4L862) is no longer hit because we are now allocating one `mp_obj_t` for each kw arg, including the ** args plus the size of the ** arg if available. So we are allocating one extra `mp_obj_t` that isn't being used for each ** arg.

dlech committed:
py/runtime: Do not overallocate when len is known.

This fixes overallocating an extra mp_obj_t when the length of *args and
**args is known. Previously we were allocating 1 mp_obj_t for each
n_args and n_kw plus the length of each *arg and **arg (if they are
known). Since n_args includes *args and n_kw includes **args, this was
allocating an extra mp_obj_t in addition to the length of these args
when unpacked.

To fix this, we just subtract 1 from the length to account for the 1
already implicitly allocated by n_args and n_kw.

Signed-off-by: David Lechner <david@pybricks.com>

dlech commented:
I have added a commit to address the overallocation issue. But it doesn't seem to affect the code coverage, so there is probably something else I am missing.

dlech committed:
py/runtime: Drop new_alloc < 4 check.

To reach this check, n_kw has to be >= 1 and therefore args2_alloc has
to be >= 2. Therefore new_alloc will always be >= 4. So this check will
never be true and can be removed.

Signed-off-by: David Lechner <david@pybricks.com>

dlech committed:
tests/basics/fun_callstardblstar: Add coverage test.

This fixes code coverage for the case where a *arg without __len__ is
unpacked and uses exactly the amount of memory that was allocated for
kw args. This triggers the code branch where the memory for the kw args
gets reallocated since it was used already by the *arg unpacking.

Signed-off-by: David Lechner <david@pybricks.com>

dlech commented:
I've added a few more commits that should take care of the dropped coverage issues. So the only question that remains is should we add a test that duplicates the cpydiff to get the last bit of coverage or should we just include all cpydiff in the coverage run?

dpgeorge commented:
> So the only question that remains is should we add a test that duplicates the cpydiff to get the last bit of coverage

Yes I think we should add another test for this in the main test suite (this is to check when there are too many positional args, right?).  Because cpydiff tests are intended to be changed/removed if/when the feature becomes compatible with CPython.  Thus cpydiff tests cannot be relied upon for coverage.

dpgeorge commented on docs/differences/python_35.rst:
```diff
@@ -8,7 +8,7 @@ Below is a list of finalised/accepted PEPs for Python 3.5 grouped into their imp
    | **Extensions to the syntax:**                                                                            | **Status**    |
    +--------------------------------------------------------+-------------------------------------------------+---------------+
-   | `PEP 448 <https://www.python.org/dev/peps/pep-0448/>`_ | additional unpacking generalizations            |               |
+   | `PEP 448 <https://www.python.org/dev/peps/pep-0448/>`_ | additional unpacking generalizations            | Completed     |
```
I don't think PEP 0448 is fully complete, we still don't support multiples in tuples, dicts etc.  I don't know if we'll ever do that.

I suggest putting "Partially done" here.

dpgeorge commented on py/runtime.c:
```diff
@@ -715,16 +714,37 @@ void mp_call_prepare_args_n_kw_var(bool have_self, size_t n_args_n_kw, const mp_
     uint args2_len = 0;
 
+    // Try to get a hint for unpacked * args length
+    int list_len = 0;
```
Now that this and `kw_dict_len` are signed, it's not clear to me that the calculations that involve these variables are guaranteed to be >= 0.... did you follow through all cases to check this?

dlech replied:
They are guaranteed to be >= 0. `list_len` can be -1 if and only if `n_args` > 0 and it is only used as `n_args + list_len`. Likewise, `kw_dict_len` can be -1 if and only if `n_kw` is > 0 and is only used as `n_kw + kw_dict_len`. Neither value can ever by less than -1 since 1 is subtracted from an unsigned value.

We could add a couple of `assert()`s to try to catch any future regressions if the code is changed.

dpgeorge replied:
`list_len` could be -2 if there are two lists, right?  Or even more negative.  But it sounds like that will still be offset by `n_args`.

Same with `kw_dict_len`?

Maybe there needs to be a test that passes in a few empty lists and empty dicts?

dlech replied:
> Neither value can ever by less than -1 since 1 is subtracted from an unsigned value.

This statement is not accurate. You are correct. The -1 is in a for loop that loops `n_args` or `n_kw` times, so `list_len` and `kw_dict_len` can be at the smallest `-n_args` and `-kw_dict_len` respectively. But is is still mathematically provable that `n_args + list_len` and `n_kw + kw_dict_len` can never be < 0.

dpgeorge replied:
I think the `int` should be `ssize_t`

dlech committed:
py/runtime: Use size_t/ssize_t instead of uint/int.

This replaces instances of uint with size_t and int with ssize_t in
the mp_call_prepare_args_n_kw_var() function since all of the variables
are used as array offsets.

Also sort headers while we are touching this.

Signed-off-by: David Lechner <david@pybricks.com>

dpgeorge commented:
Thanks for the updates.  This is looking very good now.  Merged in 1e99d29f362f8cccc60a36fcbd8590404c719f40 through 2e3f2045f9a7511987e69a8be0d3d117502e4bb1

dpgeorge commented:
I have a few extra coverage tests to add in a separate PR.
